src/psyclone/psyir/transformations/matmul2code_trans.py

```diff
@@ -183,39 +233,34 @@ def validate(self, node):
             # transformation is currently limited to Ranges which
             # specify the full extent of the dimension.
             if not vector.is_full_range(0):
-                raise GenerationError(
+                raise TransformationError(
                     "To use matmul2code_trans on matmul, index 0 of the 2nd "
                     "(vector) argument '{0}' must be a full range."
                     "".format(matrix.name))
             if len(vector.children) > 1:
-                # The 2nd index and onwards must be references.
+                # The 2nd index and onwards must not be ranges.
                 for (count, index) in enumerate(vector.children[1:]):
                     if isinstance(index, Range):
-                        raise GenerationError(
-                            "To use matmul2code_trans on matmul, indices 1 "
-                            "onwards of the second (vector) argument should "
-                            "not be Ranges but found {0} at index {1}."
-                            "".format(type(index).__name__, count+1))
+                        raise TransformationError(
+                            "To use matmul2code_trans on matmul, only the "
+                            "first index of the 2nd (vector) argument is "
+                            "permitted to be a Range but found {0} at index "
+                            "{1}.".format(type(index).__name__, count+1))
 
     def apply(self, node, options=None):
         '''Apply the MATMUL intrinsic conversion transformation to the
         specified node. This node must be a MATMUL
         BinaryOperation. Currently only the matrix vector version of
-        MATMUL is supported. If the transformation is successful then
-        an assignment which includes a MATMUL BinaryOperation node
-
-        .. code-block:: fortran
-
-            R = MATMUL(A,B)
-
-        is converted to the following equivalent inline code:
-
-        .. code-block:: fortran
-
-            do i=1,N
-                R(i) = 0.0
-                    do j=1,M
-                        R(i) = R(i) + A(i,j) * B(j)
+        MATMUL is supported. The arguments are permitted to have
+        additional dimensions (i.e. more than 2 for the matrix and
+        more than 1 for the vector) but the matrix can only have two
+        indices which are ranges and these must be the first two
+        indices and the vector can only have one index that is a range
+        and this must be the first index. Further, the ranges must be
+        for the full index space for that dimension (i.e. array slices
```

Comment:
Slightly picky but I would say that e.g. `A(:,1)` *is* an array slice. I'm not sure what e.g. `A(2:5,1)` would be called - an array subsection?

Reply:
I've just changed the offending word to subsections.